sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,156 @@
+    this.operation = operation;
+    this.stopped = stopped;
+    this.promise = new Promise(
+      (
+        resolve: (result?: TResult) => void,
+        reject: (error: PollerStoppedError | PollerCancelledError | Error) => void
+      ) => {
+        this.resolve = resolve;
+        this.reject = reject;
+      }
+    );
+    this.startPolling();
+  }
+
+  protected abstract async delay(): Promise<void>;
+
+  private async startPolling(): Promise<void> {
+    while (!this.isStopped() && !this.isDone()) {
+      await this.poll();
+      if (!this.operation.state.started) {
+        this.operation.state.started = true;
+      }
+      await this.delay();
+    }
+  }
+
+  private async pollOnce(options: { abortSignal?: AbortSignalLike } = {}): Promise<void> {
+    try {
+      if (!this.isDone()) {
+        this.operation = await this.operation.update({
+          abortSignal: options.abortSignal,
+          fireProgress: this.fireProgress.bind(this)
+        });
+        if (this.isDone() && this.resolve) {
+          this.resolve(this.operation.state.result);
+        }
+      }
+    } catch (e) {
+      this.operation.state.error = e;
+      if (this.reject) {
+        this.reject(e);
+      }
+    }
+  }
+
+  private fireProgress(properties: TProperties): void {
+    for (const callback of this.pollProgressCallbacks) {
+      callback(properties);
+    }
+  }
+
+  private async cancelOnce(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    this.operation = await this.operation.cancel(options);
+    if (this.reject) {
+      this.reject(new PollerCancelledError("Poller cancelled"));
+    }
+  }
+
+  public poll(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    if (!this.pollOncePromise) {
+      this.pollOncePromise = this.pollOnce(options);
+      const after = (): void => {
+        this.pollOncePromise = undefined;
+      };
+      this.pollOncePromise.then(after, after);
+    } else if (options.abortSignal) {
+      throw new Error("A poll request is currently pending");
+    }
+    return this.pollOncePromise;
+  }
+
+  public done(): Promise<TResult | undefined> {
+    return this.promise;
+  }
+
+  public onProgress(callback: (properties: TProperties) => void): CancelOnProgress {
+    this.pollProgressCallbacks.push(callback);
+    return (): void => {
+      this.pollProgressCallbacks = this.pollProgressCallbacks.filter((c) => c !== callback);
+    };
+  }
+
+  public isDone(): boolean {
+    return Boolean(
+      this.operation.state.completed || this.operation.state.cancelled || this.operation.state.error
+    );
+  }
+
+  public stop(): void {
+    if (!this.stopped) {
+      this.stopped = true;
+      if (this.reject) {
+        this.reject(new PollerStoppedError("Poller stopped"));
+      }
+    }
+  }
+
+  public isStopped(): boolean {
+    return this.stopped;
+  }
+
+  public cancel(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
```

Comment:
Brian ask: rename to cancelOperation:

```suggestion
  public cancelOperation(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
```